.github/workflows/ci.yml

```diff
@@ -0,0 +1,93 @@
+name: Tests
+
+on:
+  push:
+  pull_request:
+    branches:
+      - master
+
+jobs:
+  check_duplicate_runs:
+    name: Check for duplicate runs
+    continue-on-error: true
+    runs-on: ubuntu-latest
+    outputs:
+      should_skip: ${{ steps.skip_check.outputs.should_skip }}
+    steps:
+      - id: skip_check
+        uses: fkirc/skip-duplicate-actions@master
+        with:
+          concurrent_skipping: always
+          cancel_others: true
+          skip_after_successful_duplicate: true
+          paths_ignore: '["**/README.md", "**/CHANGELOG.md", "**/LICENSE"]'
+          do_not_skip: '["pull_request"]'
+
+  tests:
+    name: Run tests (Elixir ${{matrix.elixir}}, OTP ${{matrix.otp}})
+    needs: check_duplicate_runs
+    if: ${{ needs.check_duplicate_runs.outputs.should_skip != 'true' }}
+
+    strategy:
+      matrix:
+        elixir:
+        - "1.7"
+        - "1.8"
+        - "1.9"
+        - "1.10"
+        - "1.11"
+        otp:
+        - "20.3"
+        - "21.3"
+        - "22.3"
+        - "23.3"
+        exclude:
+        - elixir: "1.7"
+          otp: "23.3"
+        - elixir: "1.8"
+          otp: "23.3"
+        - elixir: "1.9"
+          otp: "23.3"
+        - elixir: "1.10"
+          otp: "20.3"
+        - elixir: "1.11"
+          otp: "20.3"
+        - elixir: "1.11"
+          otp: "21.0"
+
+    runs-on: ubuntu-latest
+
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v2
+
+    - name: Set up Elixir
+      uses: erlef/setup-elixir@v1
+      with:
+        elixir-version: ${{ matrix.elixir }}
+        otp-version: ${{ matrix.otp }}
+
+    - name: Restore deps and _build cache
+      uses: actions/cache@v2
+      with:
+        path: |
+          deps
+          _build
+        key: deps-${{ runner.os }}-${{ matrix.otp }}-${{ matrix.elixir }}-${{ hashFiles('**/mix.lock') }}-git-${{ github.sha }}
+        restore-keys: |
+          deps-${{ runner.os }}-${{ matrix.otp }}-${{ matrix.elixir }}-${{ hashFiles('**/mix.lock') }}
+          deps-${{ runner.os }}-${{ matrix.otp }}-${{ matrix.elixir }}
```

Comment:
is this line necessary? (L78 of course makes sense)

Reply:
`actions/cache` does not upload updated cache files if cache was downloaded using primary key (without fallback to restore-keys).
But sometimes it can be very useful. For example, if someone changes workflow file or adds tools like dialyzer which can update `deps` or `_build` folders content without changing `mix.lock` hash.

Adding `-git-${{ github.sha }}` to the key forces the action to upload `deps` and `_build` folders to the cache on every commit.
This does not increase cache search time or something like that.

Reply:
> Adding -git-${{ github.sha }} to the key forces the action to upload deps and _build folders to the cache on every commit. This does not increase cache search time or something like that.

Agreed there's no downside when fetching the cache but I think it would be good to not do unnecessary cache writes either. I'd personally hold off on this change until we run into a particular issue that it would have solved, but not strong feelings.

TIL we can actually remove L78 as L79 covers it, the pattern is greedy:

> `npm-foobar-d5ea0750` matches a specific hash.
> `npm-foobar-` matches cache keys prefixed with npm-foobar-.
> `npm-` matches any keys prefixed with npm-.

(https://docs.github.com/en/actions/guides/caching-dependencies-to-speed-up-workflows#matching-a-cache-key)

Reply:
Ok, removed.